rcljava/src/main/cpp/org_ros2_rcljava_RCLJava.cpp

```diff
@@ -58,20 +59,73 @@ Java_org_ros2_rcljava_RCLJava_nativeCreateNodeHandle(
 
   rcl_context_t * context = reinterpret_cast<rcl_context_t *>(context_handle);
 
+  // TODO(clalancette): we could probably make this faster by just doing these
+  // method lookups during some initialization time.  But we'd have to add a lot
+  // more infrastructure for that, and we don't expect to call
+  // 'nativeCreateNodeHandle' that often, so I think this is OK for now.
+  jclass java_util_ArrayList =
+    static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/util/ArrayList")));
+  jmethodID java_util_ArrayList_size = env->GetMethodID(java_util_ArrayList, "size", "()I");
+  jmethodID java_util_ArrayList_get = env->GetMethodID(
+    java_util_ArrayList, "get", "(I)Ljava/lang/Object;");
+
+  rcl_arguments_t arguments = rcl_get_zero_initialized_arguments();
+  char ** argv = NULL;
+  jint argc = env->CallIntMethod(cli_args, java_util_ArrayList_size);
+  argv = static_cast<char **>(malloc(argc * sizeof(char *)));
+  for (jint i = 0; i < argc; ++i) {
+    jstring element =
+      static_cast<jstring>(env->CallObjectMethod(cli_args, java_util_ArrayList_get, i));
+    argv[i] = const_cast<char *>(env->GetStringUTFChars(element, nullptr));
+  }
+  rcl_ret_t ret = rcl_parse_arguments(argc, argv, rcl_get_default_allocator(), &arguments);
+  for (jint i = 0; i < argc; ++i) {
+    jstring element =
+      static_cast<jstring>(env->CallObjectMethod(cli_args, java_util_ArrayList_get, i));
+    env->ReleaseStringUTFChars(element, argv[i]);
+  }
+  free(argv);
+  if (ret != RCL_RET_OK) {
+    std::string msg = "Failed to parse node arguments: " + std::string(rcl_get_error_string().str);
+    rcl_reset_error();
+    rcljava_throw_rclexception(env, ret, msg);
+    return 0;
+  }
+
   rcl_node_t * node = static_cast<rcl_node_t *>(malloc(sizeof(rcl_node_t)));
   *node = rcl_get_zero_initialized_node();
 
-  rcl_node_options_t default_options = rcl_node_get_default_options();
-  rcl_ret_t ret = rcl_node_init(
-    node, node_name.c_str(), namespace_.c_str(), context, &default_options);
+  rcl_node_options_t options = rcl_node_get_default_options();
+  options.use_global_arguments = use_global_arguments;
+  options.arguments = arguments;
+  options.enable_rosout = enable_rosout;
+  ret = rcl_node_init(node, node_name.c_str(), namespace_.c_str(), context, &options);
   if (ret != RCL_RET_OK) {
     std::string msg = "Failed to create node: " + std::string(rcl_get_error_string().str);
     rcl_reset_error();
+    free(node);
+    if (rcl_arguments_fini(&arguments) != RCL_RET_OK) {
+      // We are already throwing an exception, just ignore the return here
+    }
```

Comment:
```suggestion
    // We are already throwing an exception, just ignore the return here
    (void)rcl_arguments_fini(&arguments);
```

Reply:
Oh, except that doesn't work.  Since the function is marked `warn_unused_result`, the cast to void doesn't work:

```
/home/ubuntu/rcljava_ws/src/ros2_java/ros2_java/rcljava/src/main/cpp/org_ros2_rcljava_RCLJava.cpp: In function ‘jlong Java_org_ros2_rcljava_RCLJava_nativeCreateNodeHandle(JNIEnv*, jclass, jstring, jstring, jlong, jobject, jboolean, jboolean)’:
/home/ubuntu/rcljava_ws/src/ros2_java/ros2_java/rcljava/src/main/cpp/org_ros2_rcljava_RCLJava.cpp:106:5: warning: ignoring return value of ‘rcl_ret_t rcl_arguments_fini(rcl_arguments_t*)’, declared with attribute warn_unused_result [-Wunused-result]
     (void)rcl_arguments_fini(&arguments);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
That's surprising ....
I think that:

```cpp
rcl_ret_t ret = rcl_arguments_fini(&arguments);
(void)ret;
```

works, but it's not much cleaner.

Reply:
Yeah, that version works.  I personally prefer the version already there for slightly aesthetic reasons, so unless you feel strongly about it I'm just going to leave it.